lib/config_cat.ex

```diff
@@ -0,0 +1,187 @@
+defmodule ConfigCat do
+  use GenServer
+
+  alias ConfigCat.{FetchPolicy, Rollout}
+  alias HTTPoison.Response
+
+  require Logger
+
+  @base_url "https://cdn.configcat.com"
+  @base_path "configuration-files"
+  @config_filename "config_v4.json"
+
+  def start_link(sdk_key, options \\ [])
+
+  def start_link(nil, _options), do: {:error, :missing_sdk_key}
+
+  def start_link(sdk_key, options) do
+    with {name, options} <- Keyword.pop(options, :name, __MODULE__),
+         {initial_config, options} <- Keyword.pop(options, :initial_config) do
+      initial_state = %{
+        config: initial_config,
+        etag: nil,
+        last_update: nil,
+        options: Keyword.merge(default_options(), options),
+        sdk_key: sdk_key
+      }
+
+      GenServer.start_link(__MODULE__, initial_state, name: name)
+    end
+  end
+
+  defp default_options, do: [api: ConfigCat.API, fetch_policy: FetchPolicy.auto()]
+
+  def get_value(key, default_value, user_or_options \\ []) do
+    if Keyword.keyword?(user_or_options) do
+      get_value(key, default_value, nil, user_or_options)
+    else
+      get_value(key, default_value, user_or_options, [])
+    end
+  end
+
+  def get_value(key, default_value, user, options) do
+    client = Keyword.get(options, :client, __MODULE__)
+    GenServer.call(client, {:get_value, key, default_value, user})
+  end
+
+  def force_refresh(client \\ __MODULE__) do
+    GenServer.call(client, :force_refresh)
+  end
+
+  @impl GenServer
+  def init(state) do
+    {:ok, state, {:continue, :maybe_init_fetch}}
+  end
+
+  @impl GenServer
+  def handle_call({:get_value, key, default_value, user}, _from, state) do
+    with {:ok, new_state} <- maybe_refresh(state),
+         value <- Rollout.evaluate(key, user, default_value, new_state.config) do
+      {:reply, value, new_state}
+    else
+      error -> {:reply, error, state}
+    end
+  end
+
+  @impl GenServer
+  def handle_call(:force_refresh, _from, state) do
+    with {:ok, new_state} <- refresh(state) do
+      {:reply, :ok, new_state}
+    else
+      error -> {:reply, error, state}
+    end
+  end
+
+  defp schedule_initial_fetch?(%{options: options}) do
```

Comment:
Please extract all of the refresh logic to a config_fetcher.ex

Reply:
Also please extract the caching to a cache.ex and provide ability to override it with the options

Reply:
Extracted the first comment to #14. The second comment is already covered by #10.